### PR TITLE
[util] Replace random.py with secure_prng.py

### PR DESCRIPTION
**Pull request by vrozic**
This PR aims to replace the python pseudo-random number generator used for generating netlist constants and scrambling keys.
This PR consists of four commits:

1. Adds an implementation of a class `secure_prng.py` which is based on NIST CTR_DRBG generator. In addition, this commit contains unit tests with official test vectors obtained from [https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/random-number-generators#DRBG](https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/random-number-generators#DRBG)
2. Replaces calls to `random.py` methods with calls to `secure_prng.py` methods.
3. Replaces 64 bit seeds in hjson files with 256 bit seeds.
4. Updates auto-generated files.

### CODE REVIEW

File: util/topgen/drg3.py
Comment (vogelpi) [inline util/topgen/drg3.py:55]: This should actually be done the instantiate function. Is it possible to initialize these values to something like `null` here?

**Review by vogelpi — COMMENTED** [review]
Thanks for the PR @vrozic and the good documentation / clear references to the spec.

This looks very good! I have just a few things. Whether all the byte orderings and conversions are correct we can only assess using test vectors which can e.g. be found here:
https://csrc.nist.gov/projects/cryptographic-algorithm-validation-program/random-number-generators#DRBG

File: util/topgen/drg3.py
Comment (vrozic) [reply util/topgen/drg3.py:66]: Good point.
I've just finished some preliminary testing using the official test vector and everything seems to be in order.
I will extend the class with the unit test so that it is easier to verify.

**Comment by vrozic** [comment]
In the latest push I've added the unit_test() method to make verification easier.
The test uses a 256 bit entropy input, makes two requests to CTR_DRBG and prints out the returned bits as well as the internal state.

For example, you can run the test as:
```
entropy_input = 0x14a9a1e20ca3686f3ec09603f6730bdd73a9cc8c32388ec00bc28e6c9ec65d87
drg3.unit_test(entropy_input)
```

and the corresponding output is:
```
Entropy Input =  0x14a9a1e20ca3686f3ec09603f6730bdd73a9cc8c32388ec00bc28e6c9ec65d87
INSTANTIATE: 
Key =  0x4c4b5d2cf6dd580e08bf8b5452944e87
V   =  0x70211642528e2d52f8ea4cd5ef74a3ff

GENERATE(FIRST CALL):
Key =  0x804b3b0d3d7a83bda3700f5dfc34041c
V   =  0x4fdaf404e9e669c78263df43410bf80a

ReturnedBits =  0xc12946457ad89dde788e13c5092bbf3e9c7f61eac73261d01d396a1d9d1163eee847a561d8e52080cd9bdce570141782ebf167c4700a66aca6fd5c9116a7be49
GENERATE(SECOND CALL):
Key =  0x5aa43bd012bcb092e7b39c9427af2b7f
V   =  0xe54c4cf6d443b47862923c435148c17b
```

For comparison, the expected output from the test vectors file is:
```
EntropyInput = 14a9a1e20ca3686f3ec09603f6730bdd73a9cc8c32388ec00bc28e6c9ec65d87
Nonce = 
PersonalizationString = 
** INSTANTIATE:
	Key = 4c4b5d2cf6dd580e08bf8b5452944e87
	V   = 70211642528e2d52f8ea4cd5ef74a3ff

AdditionalInput = 
** GENERATE (FIRST CALL):
	Key = 804b3b0d3d7a83bda3700f5dfc34041c
	V   = 4fdaf404e9e669c78263df43410bf80a

AdditionalInput = 
ReturnedBits = c12946457ad89dde788e13c5092bbf3e9c7f61eac73261d01d396a1d9d1163eee847a561d8e52080cd9bdce570141782ebf167c4700a66aca6fd5c9116a7be49
** GENERATE (SECOND CALL):
	Key = 5aa43bd012bcb092e7b39c9427af2b7f
	V   = e54c4cf6d443b47862923c435148c17b
```

I've tested several vectors for our use case
```
[AES-128 no df]
[PredictionResistance = False]
[EntropyInputLen = 256]
[NonceLen = 0]
[PersonalizationStringLen = 0]
[AdditionalInputLen = 0]
[ReturnedBitsLen = 512]
```
 and outputs as well as states always match the expectation. Therefore, I think that all byte-ordering and data manipulation is working correctly.

File: util/topgen/drg3.py
Comment (vrozic) [reply util/topgen/drg3.py]: Good point. I've changed all integer values to None and replaced update function with `drg3.reseed(0)`.

File: util/topgen/drg3.py
Comment (vrozic) [reply util/topgen/drg3.py]: Originally I've listed all CTR_DRBG parameters as variables to make it easier to follow specification.
In the latest push I've removed some of the unused variables.
`seedlen` and `ctr_len` are not used in the code but I'm refering to them in the comments, so I've converted these lines into comments. I think it makes it easier to match the code to specification.

**Review by vogelpi — COMMENTED** [review]
This looks mostly good, thanks @vrozic for working on this!

I have some questions regarding extending the unit test to cover more functionality and around the initialization.

File: util/topgen/secure_prng.py
Comment (marnovandermaas) [inline util/topgen/secure_prng.py:28]: We should probably remove this commented out code.

File: util/topgen/secure_prng.py
Comment (marnovandermaas) [inline util/topgen/secure_prng.py:23]: Can we think of a better name for this? Maybe call it `ctr_drbg_no_input_256` or some variant of that?

File: util/design/lib/OtpMemMap.py
Comment (marnovandermaas) [inline util/design/lib/OtpMemMap.py:13]: Is there any way to remove this `strong_random` import?

File: util/topgen.py
Comment (marnovandermaas) [inline util/topgen.py:33]: Do we still need `strong_random` here?

File: util/topgen/BUILD
Comment (marnovandermaas) [inline util/topgen/BUILD:23]: Is this already in the python requirements?

File: util/topgen.py
Comment (marnovandermaas) [reply util/topgen.py:1298]: You could also move the line with `-o` to align with `--rnd_cnst_seed`.

File: hw/ip/lc_ctrl/data/lc_ctrl_state.hjson
Comment (marnovandermaas) [inline hw/ip/lc_ctrl/data/lc_ctrl_state.hjson:12]: How are these seeds generated?

**Review by marnovandermaas — COMMENTED** [review]
Thanks for creating this PR, some initial comments from my side.

File: hw/ip/lc_ctrl/data/lc_ctrl_state.hjson
Comment (vrozic) [reply hw/ip/lc_ctrl/data/lc_ctrl_state.hjson:12]: I've used `secrets.py` module which uses the entropy source from the operating system.

File: util/topgen/BUILD
Comment (vrozic) [reply util/topgen/BUILD:23]: yes
`pycryptodome==3.15.0`

File: util/topgen.py
Comment (vrozic) [reply util/topgen.py:33]: This is still an open discussion point, but the probable answer is no.
i plan to do `strong_random` cleanup in a separate PR

File: hw/ip/lc_ctrl/data/lc_ctrl_state.hjson
Comment (marnovandermaas) [reply hw/ip/lc_ctrl/data/lc_ctrl_state.hjson:12]: Maybe you can add a comment to this file saying how you've generated this (i.e. what Python commands did you use).

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: Thanks for catching this. It's corrected now.

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: Sounds good. I've changed it now to error out if a user attempts to get an output without reseeding PRNG first.
Originally, I've had the default reseed to mimic what random.py does.

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: For now, I've replaced calls to `CTR_DRBG_Generate(520)` with calls to `getrandbits(512)`.
I'll think about making a more elaborate test for `getrandbits()`

File: util/topgen.py
Comment (vrozic) [reply util/topgen.py]: For now I've fixed alignment and moved `{seed}` to the next line to avoid lint failures

File: hw/ip/lc_ctrl/data/lc_ctrl_state.hjson
Comment (vrozic) [reply hw/ip/lc_ctrl/data/lc_ctrl_state.hjson:13]: Good idea. I've added comments to hjson files

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: This is intentional.
The test vectors make two calls to `CTR_DRBG_Generate`, and only output the numbers of the second call. So we don't need the `ReturnedBits` after the first call because we have nothing to compare them with.

File: util/topgen/secure_prng.py
Comment (vogelpi) [reply util/topgen/secure_prng.py:23]: I would keep the name. IMHO it's descriptive enough and the implementation provides many comments to detail what construction exactly is used under the hood.

File: util/topgen/secure_prng.py
Comment (vogelpi) [reply util/topgen/secure_prng.py:28]: FYI: there have been previous discussions on this before `seedlen` and `ctr_len` are the terms used in the NIST spec. Due to the simplified / stripped-down implementation we've selected here, they are not used in the code, but @vrozic 's code comments refer to them. That's why it was decided to keep the variables but comment them. @vrozic it probably makes sense to add a comment on this here. The same question might come up again. This is already the second time ;-)

File: util/topgen/secure_prng.py
Comment (vogelpi) [reply util/topgen/secure_prng.py:168]: Ping

File: util/topgen/secure_prng.py
Comment (vogelpi) [reply util/topgen/secure_prng.py]: Okay thanks for clarifying.

Comment (vogelpi) [reply util/topgen/secure_prng.py]: You could for example use `random` / any other PRNG to randomize the number of byte per `getrandbits()` call.

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: > We should probably remove this commented out code.

I've added some text to explain it.

File: util/topgen/secure_prng.py
Comment (jadephilipoom) [inline util/topgen/secure_prng.py:189]: Is there any expected control flow where we reseed the PRNG with the wrong length of bits? I'd expect that in this case we'd exit with an error, not just a warning, since it's a security-critical operation. It's probably okay to take bytes here rather than an integer, which makes the length much easier to check.
```suggestion
    def reseed(self, seed):
        """Re-seed PRNG

        Instantiates CTR_DRBG with a 256-bit entropy_input.
        Empties returned_bits list in case that it contains residual numbers
        from previous instantiation.
        """
        if int.from_bytes(seed, byteorder='big') == 0:
            log.warning('WARNING: PRNG seeded with 0.')
        if len(seed)  != 32:
            log.error('Seed must be 32 bytes long.')
            sys.exit(1)
        entropy_input = seed
```

File: util/topgen/secure_prng.py
Comment (jadephilipoom) [inline util/topgen/secure_prng.py:116]:
```suggestion
            self.V = (self.V + 1) % (1 << self.blocklen)
```

File: util/topgen/secure_prng.py
Comment (jadephilipoom) [inline util/topgen/secure_prng.py:69]: These need to wrap around the block length:
```suggestion
        pt0 = (self.V + 1) % (1 << self.blocklen)
        pt1 = (self.V + 2) % (1 << self.blocklen)
```

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py:85]: Yes indeed. I've added the check now.
This was one of the checks that I didn't implement originally because we only call `CTR_DRBG_Generate` through `fetchbyte` which only makes requests for 512 bits. But it certainly doesn't hurt adding it now.

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: Thanks for catching this. Corrected.

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: Good catch. Corrected.

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: > Is there any expected control flow where we reseed the PRNG with the wrong length of bits?

Unfortunately there is. Dvsim by default uses `--build-seed` of only 32 bits. Also, some people might run simulations with randomly typing seed parameters of arbitrary length.
I'm also not exactly sure how we choose the seed for the nightly regressions.
I guess that the correct thing to do would be to modify dvsim to check/expand the seed and to always make reseed calls with 256 bit. For the time being I've added a TO DO comment in the code and I'll open an issue to track this.

File: hw/ip/lc_ctrl/data/lc_ctrl_state.hjson
Comment (timothytrippel) [reply hw/ip/lc_ctrl/data/lc_ctrl_state.hjson:13]: Yes, there is a TO checklist that includes changing this seed (and the others) before generating the SV files used for final TO. For ES, I believe it was a bit manual with some scripting, but once this change goes in, I aim to refactor how our tooling (i.e., bazel, dvsim) drives the scripts that ingest these seeds to make things a bit more seamless.

**Review by jadephilipoom — APPROVED** [review]

**Comment by vrozic** [comment]
For visibility, RFC has been approved #19228

File: util/topgen/secure_prng.py
Comment (vogelpi) [reply util/topgen/secure_prng.py]: Thanks @vrozic ! Now I understand how this works. It wasn't clear to me how Python internally represents integer and string objects and how to convert the lengths. This makes sense to me now.

File: util/topgen/secure_prng_test.py
Comment (vogelpi) [inline util/topgen/secure_prng_test.py:289]: Smart :-)

File: util/topgen/secure_prng.py
Comment (vogelpi) [inline util/topgen/secure_prng.py:307]: I am wondering if this function should better be moved to `secure_prng_test.py` as all the other test code.

File: util/topgen/secure_prng_test.py
Comment (vogelpi) [inline util/topgen/secure_prng_test.py:1]: This file should get the executable flag set, i.e., `chmod a+x secure_prng_test.py`.

**Review by vogelpi — APPROVED** [review]
Great work @vrozic , thanks for addressing the remaining comments, LGTM!

I believe the last 3 commits should be combined for merging (thanks for splitting the auto-gen files for review though!). Otherwise they're not atomic: reverting one commit will leave the repo in a broken state.

Maybe you can also create an issue to remove strong_random after merging this. Such that we don't forget. Thanks for doing this work!

File: util/topgen/secure_prng.py
Comment (vogelpi) [reply util/topgen/secure_prng.py]: Fantastic!

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py:307]: Unfortunately, this function needs to read instance variables `_inst.Key` and `_inst.V`. I would rather that the test continues using protected instance `_inst` because this is how the module will be used in practice.

**Comment by vogelpi** [comment]
CHANGE AUTHORIZED:
hw/ip/otp_ctrl/data/otp_ctrl_img_dev.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_prod.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_raw.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_rma.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_locked0.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_locked1.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked0.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked1.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked2.hjson
hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson
 hw/ip/otp_ctrl/rtl/otp_ctrl_part_pkg.sv
hw/top_earlgrey/data/top_earlgrey.hjson

File: util/topgen/secure_prng.py
Comment (vrozic) [reply util/topgen/secure_prng.py]: OK, I've opened #19434 to track this and changed the comment.

File: util/topgen.py
Comment (vrozic) [reply util/topgen.py:33]: I've opened #19433 to track this.

**Comment by msfschaffner** [comment]
CHANGE AUTHORIZED:
hw/ip/otp_ctrl/data/otp_ctrl_img_dev.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_prod.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_raw.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_rma.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_locked0.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_locked1.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked0.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked1.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked2.hjson
hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson
hw/ip/otp_ctrl/rtl/otp_ctrl_part_pkg.sv
hw/top_earlgrey/data/top_earlgrey.hjson

**Review by msfschaffner — APPROVED** [review]
Thanks for the PR @vrozic! This LGTM, modulo the pending cleanups / squashes necessary to make this atomic.

Can you also cherry-pick this set of changes onto the integrated branch after merging?

Thanks!

**Comment by andreaskurth** [comment]
CHANGE AUTHORIZED:
hw/ip/lc_ctrl/rtl/lc_ctrl_state_pkg.sv
hw/ip/otp_ctrl/data/otp_ctrl_img_dev.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_prod.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_raw.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_rma.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_locked0.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_locked1.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked0.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked1.hjson
hw/ip/otp_ctrl/data/otp_ctrl_img_test_unlocked2.hjson
hw/ip/otp_ctrl/data/otp_ctrl_mmap.hjson
hw/ip/otp_ctrl/rtl/otp_ctrl_part_pkg.sv
hw/top_earlgrey/data/top_earlgrey.hjson

**Comment by vrozic** [comment]
Thanks @marnovandermaas,  @vogelpi , @jadephilipoom, @timothytrippel and @msfschaffner for your reviews and helpful comments.
I've now made all requested changes.
The latest push squashes the last three commits.
In my view, this is ready to merge.

**Pull request merged**
